File: Baekjoon Online Judge/easy/01463.c
#include <stdio.h>
#define SIZE 1000001

int main()
{
    // Init
    int N;
    int dp[SIZE];
    
    // Input
    scanf("%d", &N);
    
    // Solve
    for (int i = 2; i < SIZE; ++i) {
        dp[i] = dp[i - 1] + 1;
        if (i % 2 == 0 && dp[i / 2] + 1 < dp[i])
            dp[i] = dp[i / 2] + 1;
        if (i % 3 == 0 && dp[i / 3] + 1 < dp[i])
            dp[i] = dp[i / 3] + 1;
    }
    
    // Output
    printf("%d", dp[N]);
}

File: baekjoon-online-judge/easy/01463.py
from collections import deque
from sys import stdin

input = stdin.readline

if __name__ == '__main__':
    # Input
    X = int(input())

    # Solve
    queue = deque([(X, 0)])
    while queue:
        x, cnt = queue.popleft()
        if x == 1:
            # Output
            print(cnt)
            break
        if x % 3 == 0:
            queue.append((x // 3, cnt + 1))
        if x % 2 == 0:
            queue.append((x // 2, cnt + 1))
        queue.append((x - 1, cnt + 1))
